### PR TITLE
Docstrings for image classification data

### PR DESCRIPTION
**Pull request by ethanwharris**
## What does this PR do?

Adds docstrings and doctests to core dataloading methods in ImageClassificationData

## Before submitting
- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/lightning-flash/tree/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? [not needed for typos/docs]
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-flash/blob/master/CHANGELOG.md)?

## PR review
 - [x] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: flash/image/classification/data.py [hunk 1]
```diff
@@ -207,6 +471,101 @@ def from_data_frame(
...
+        Input images will be extracted from the ``input_field`` in the DataFrame.
+        The supported file extensions are: ``.jpg``, ``.jpeg``, ``.png``, ``.ppm``, ``.bmp``, ``.pgm``, ``.tif``,
+        ``.tiff``, ``.webp``, and ``.npy``.
+        The targets will be extracted from the ``target_fields`` in the DataFrame and can be in any of our
+        :ref:`supported classification target formats <formatting_classification_targets>`.
+        To learn how to customize the transforms applied for each stage, read our
+        :ref:`customizing transforms guide <customizing_transforms>`.
+
+        Args:
+            input_field: The field in the DataFrames containing the image files.
```
Comment (Borda) [inline flash/image/classification/data.py:486]:
```suggestion
            input_field: The field in the DataFrames containing the name of image files.
```

File: flash/image/classification/data.py [hunk 2]
```diff
@@ -207,6 +471,101 @@ def from_data_frame(
...
+            target_fields: The field or list of fields in the DataFrames containing the targets.
```
Comment (Borda) [inline flash/image/classification/data.py:487]: what the field means, is it int, string label?

File: flash/image/classification/data.py [hunk 3]
```diff
@@ -207,6 +471,101 @@ def from_data_frame(
...
+            train_data_frame: The pandas DataFrame to use when training.
+            train_images_root: The root directory containing train images.
```
Comment (Borda) [inline flash/image/classification/data.py:489]:
```suggestion
            train_images_root_dir: The root directory containing train images.
```

File: flash/image/classification/data.py [hunk 4]
```diff
@@ -207,6 +471,101 @@ def from_data_frame(
...
+            train_resolver: Optionally provide a function which converts an entry from the ``input_field`` into an image
```
Comment (Borda) [inline flash/image/classification/data.py:490]: that is quite redundant, just updating the dataframe, but rather allowing custom image loading, can be handy for example for medical images such as DICON

File: flash/image/classification/data.py [hunk 5]
```diff
@@ -207,6 +471,101 @@ def from_data_frame(
...
+            predict_data_frame: The pandas DataFrame to use when predicting.
+            predict_images_root: The root directory containing predict images.
+            predict_resolver: Optionally provide a function which converts an entry from the ``input_field`` into an
+                image file path.
+            train_transform: The :class:`~flash.core.data.io.input_transform.InputTransform` type to use when training.
+            val_transform: The :class:`~flash.core.data.io.input_transform.InputTransform` type to use when validating.
+            test_transform: The :class:`~flash.core.data.io.input_transform.InputTransform` type to use when testing.
+            predict_transform: The :class:`~flash.core.data.io.input_transform.InputTransform` type to use when
+              predicting.
+            input_cls: The :class:`~flash.core.data.io.input.Input` type to use for loading the data.
```
Comment (Borda) [inline flash/image/classification/data.py:509]: without click cross-link, no idea what it means

File: flash/image/classification/data.py [hunk 6]
```diff
@@ -134,6 +290,60 @@ def from_numpy(
...
+                ...     predict_data=[np.random.rand(3, 64, 64)],
+                ...     transform_kwargs=dict(image_size=(128, 128)),
+                ...     batch_size=2,
+                ... )
+                >>> datamodule.num_classes
+                2
+                >>> datamodule.labels
+                ['cat', 'dog']
+                >>> model = ImageClassifier(backbone="resnet18", num_classes=datamodule.num_classes)
+                >>> trainer = Trainer(limit_train_batches=1, max_epochs=1)
```
Comment (tchaton) [inline flash/image/classification/data.py:341]:
```suggestion
                >>> trainer = Trainer(fast_dev_run=True)
```

File: flash/image/classification/data.py [hunk 7]
```diff
@@ -252,6 +611,114 @@ def from_csv(
         transform_kwargs: Optional[Dict] = None,
         **data_module_kwargs: Any,
     ) -> "ImageClassificationData":
+        """Load the :class:`~flash.image.classification.data.ImageClassificationData` from CSV files containing
```
Comment (Borda) [inline flash/image/classification/data.py:614]: maybe explicitly mention that is parallel to `from_data_frame` just instead of data frame it is CSV

File: flash/image/classification/data.py [hunk 8]
```diff
@@ -252,6 +611,114 @@ def from_csv(
...
+            transform_kwargs: Dict of keyword arguments to be provided when instantiating the transforms.
+            data_module_kwargs: Additional keyword arguments to provide to the
+              :class:`~flash.core.data.data_module.DataModule` constructor.
+
+        Returns:
+            The constructed :class:`~flash.image.classification.data.ImageClassificationData`.
+
+        Example::
+
+            .. testsetup::
```
Comment (tchaton) [inline flash/image/classification/data.py:659]: Could this be added to a private file _data_doc.py as this could be used to generate more complex data.

File: flash/image/classification/data.py (hunk 7 above)
Comment (Borda) [reply flash/image/classification/data.py:614]: said so, why have this function at all when you do the same with
```py
.from_data_frame(..., train_data_frame=pd.read_csv(train_csv), ...)
```

**Review by tchaton — APPROVED** [review]
LGTM ! Awesome work !

File: flash/image/classification/data.py (hunk 8 above)
Comment (tchaton) [reply flash/image/classification/data.py:659]: >>> from flash.image.classification.data_doc import test_from_folders
>>> test_from_folders()

Just an idea :)

File: flash/image/classification/data.py [hunk 9]
```diff
@@ -71,6 +76,71 @@ def from_files(
...
+
+                >>> from PIL import Image
+                >>> rand_image = Image.fromarray(np.random.randint(0, 255, (64, 64, 3), dtype="uint8"))
+                >>> rand_image.save("image_1.png")
+                >>> rand_image.save("image_2.png")
+                >>> rand_image.save("image_3.png")
+                >>> rand_image.save("predict_image_1.png")
+                >>> rand_image.save("predict_image_2.png")
+
+            .. doctest::
```
[the remark this answers is not in the document]
Comment (ethanwharris) [reply flash/image/classification/data.py:122]: See other reply

File: .pre-commit-config.yaml [hunk 10]
```diff
@@ -69,7 +69,7 @@ repos:
     rev: v1.11.0
     hooks:
       - id: blacken-docs
-        args: [ --line-length=120 ]
+        args: [ --line-length=120, --skip-errors ]
```
[the remark this answers is not in the document]
Comment (ethanwharris) [reply .pre-commit-config.yaml:72]: There is a bug in blacken-docs where it gives a pre-commit error incorrectly with lines starting with `>>>`. This just makes it still run but without failing the pre-commit check.

File: docs/source/general/classification_targets.rst [hunk 11]
```diff
@@ -0,0 +1,7 @@
+.. _formatting_classification_targets:
```
[the remark this answers is not in the document]
Comment (ethanwharris) [reply docs/source/general/classification_targets.rst:1]: Yeah, we could auto-tag, it would standardise the labels which would be nice

File: flash/image/classification/data.py (hunk 8 above)
Comment (ethanwharris) [reply flash/image/classification/data.py:659]: Yeah, we could do this, I think eventually they will come from `tests/helpers`. It doesn't matter too much as the contents of the `testsetup` block will be hidden in the rendered docs so never seen by users.

File: flash/image/classification/data.py (hunk 7 above)
Comment (ethanwharris) [reply flash/image/classification/data.py:614]: > said so, why have this function at all when you do the same with

Just for ease of use. It will also automatically set the root to be the root of the CSV file which can be useful. I'll add something to the docs about it.

File: flash/image/classification/data.py [hunk 12]
```diff
@@ -71,6 +76,71 @@ def from_files(
...
+
+        Example::
+
+            .. testsetup::
+
+                >>> from PIL import Image
+                >>> rand_image = Image.fromarray(np.random.randint(0, 255, (64, 64, 3), dtype="uint8"))
+                >>> rand_image.save("image_1.png")
+                >>> rand_image.save("image_2.png")
+                >>> rand_image.save("image_3.png")
```
[the remark this answers is not in the document]
Comment (ethanwharris) [reply flash/image/classification/data.py:118]: Done :smiley:

File: flash/image/classification/data.py (hunk 6 above)
Comment (ethanwharris) [reply flash/image/classification/data.py:341]: Done :smiley:

File: flash/image/classification/data.py (hunk 4 above)
Comment (ethanwharris) [reply flash/image/classification/data.py:490]: Perhaps not needed for dataframe but useful for CSV. Will leave it for now. Custom image loading would definitely be a good thing to have...

File: flash/image/classification/data.py (hunk 1 above)
Comment (ethanwharris) [reply flash/image/classification/data.py:486]: Done :smiley:

File: flash/image/classification/data.py (hunk 2 above)
Comment (ethanwharris) [reply flash/image/classification/data.py:487]: Clarified :smiley:

File: flash/image/classification/data.py (hunk 3 above)
Comment (ethanwharris) [reply flash/image/classification/data.py:484]: We could change the name, I don't have any strong feelings about it. Will leave it out of this PR though to avoid breaking changes.

File: flash/image/classification/data.py (hunk 5 above)
Comment (ethanwharris) [reply flash/image/classification/data.py:504]: Yeah, would be good to add a page on how to use the `input_cls` argument for customizing data loading. Will do in a follow-up :smiley:
